### PR TITLE
TypeResolutionTable

### PR DESCRIPTION
**Pull request by moodmosaic**
As described in #91.

### CODE REVIEW

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        public void VerifyGuardClauses(GuardClauseAssertion assertion)
+        {
+            assertion.Verify(typeof(TypeResolutionTable));
+        }
+
+        [Theory, AutoAtomData]
+        public void EntriesIsCorrectWhenInitializedWithArray(
+            TypeResolutionEntry[] expected)
+        {
+            var sut = new TypeResolutionTable(expected);
```
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Did you know that, instead of using the constructor in the test, you can ask for it in the method arguments, and use the `[FavorArrays]` attribute?

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            var actual = sut.Entries;
+            Assert.True(expected.SequenceEqual(actual));
+        }
+
+        [Theory, AutoAtomData]
+        public void EntriesIsCorrectWhenInitializedWithEnumerable(
+            IEnumerable<TypeResolutionEntry> expected)
+        {
+            var sut = new TypeResolutionTable(expected);
```
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Did you know that, instead of using the constructor in the test, you can ask for it in the method arguments, and use the `[FavorEnumerables]` attribute?

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            Assert.True(expected.SequenceEqual(actual));
+        }
+
+        [Theory, AutoAtomData]
+        public void ResolveReturnsCorrectResult(
+            TypeResolutionEntry[] entries)
+        {
+            var entry = entries.PickRandom();
+            var expected = entry.Resolution;
+            var sut = new TypeResolutionTable(entries);
```
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Why not ask for the `sut` from the method arguments, and then select one from `sut.Entries`?

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            var actual = sut.Resolve(entry.LocalName, entry.XmlNamespace);
+
+            Assert.Equal(expected, actual);
+        }
+
+        [Theory, AutoAtomData]
+        public void ResolveThrowsWhenInputCanNotBeMappedToProperType(
+            TypeResolutionEntry notMapped,
+            TypeResolutionEntry[] entries)
+        {
```
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs:64]: Do we positively, for sure, know that `notMapped` isn't in `entries`?

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            var sut = new TypeResolutionTable(entries);
```
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Why not ask for `sut` from the method arguments?

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            Assert.Throws<ArgumentException>(() =>
+                sut.Resolve(notMapped.LocalName, notMapped.XmlNamespace));
+        }
+    }
```
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs:100]: Is there a test case where we ask for an unmapped local name, but a mapped namespace; or for a mapped local name, but an unmapped namespace?

File: AtomEventStore/TypeResolutionEntry.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        /// Gets the <see cref="Type"/> resolved by the XML namespace and the
+        /// local name of the <see cref="TypeResolutionEntry"/>.
+        /// </summary>
+        /// <value>
+        /// The <see cref="Type"/> resolved by the XML namespace and the local
+        /// name of the <see cref="TypeResolutionEntry"/>, as originally
+        /// provided via the constructor.
+        /// </value>
+        /// <seealso cref="TypeResolutionEntry(string, string, Type)" />
+        public Type Resolution
```
Comment (ploeh) [inline AtomEventStore/TypeResolutionEntry.cs]: How about `Type` or `ResolvedType`?

([Resolution](http://dictionary.reference.com/browse/resolution) is quite an overloaded noun.)

File: AtomEventStore/TypeResolutionTable.cs [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        /// <para>
+        /// The constructor arguments are subsequently available on the object
+        /// as properties.
+        /// </para>
+        /// </remarks>
+        /// <exception cref="System.ArgumentNullException">
+        /// <paramref name="entries" /> is <see langword="null" />.
+        /// </exception>
+        public TypeResolutionTable(IEnumerable<TypeResolutionEntry> entries)
+            : this(entries.ToArray())
```
Comment (ploeh) [inline AtomEventStore/TypeResolutionTable.cs:33]: What happens if the client passes in an infinite sequence?

File: AtomEventStore/TypeResolutionTable.cs [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+            try
+            {
+                return this.entries
+                    .Single(x =>
+                        x.LocalName == localName &&
+                        x.XmlNamespace == xmlNamespace)
+                    .Resolution;
+            }
+            catch (InvalidOperationException e)
```
Comment (ploeh) [inline AtomEventStore/TypeResolutionTable.cs]: Is exception handling a good way to model control flow?

Comment (moodmosaic) [inline AtomEventStore/TypeResolutionTable.cs]: I was thinking to return an empty sequence, or return a `Maybe`, but then I read the [protocol](https://github.com/GreanTech/AtomEventStore/blob/ad1064b38dc2871e2b6f4a9fbca03589089547b5/AtomEventStore/ITypeResolver.cs#L35-37) of `ITypeResolver`.

File: AtomEventStore/TypeResolutionTable.cs (hunk 8 above)
Comment (moodmosaic) [inline AtomEventStore/TypeResolutionTable.cs:33]: OMG, I've totally missed that, and you also wrote [about it](http://blog.ploeh.dk/2013/07/20/linq-versus-the-lsp/) in the past. How about `IReadOnlyCollection<T>`?

File: AtomEventStore/TypeResolutionEntry.cs (hunk 7 above)
Comment (moodmosaic) [inline AtomEventStore/TypeResolutionEntry.cs]: I tend to prefer `ResolvedType` as it's most specific than just `Type`. What would you prefer?

File: AtomEventStore/TypeResolutionTable.cs (hunk 9 above)
Comment (ploeh) [inline AtomEventStore/TypeResolutionTable.cs]: The method, as implemented here, has the correct behaviour. Could it be _implemented_ without relying on `try`/`catch`?

File: AtomEventStore/TypeResolutionTable.cs (hunk 8 above)
Comment (ploeh) [inline AtomEventStore/TypeResolutionTable.cs:33]: `IReadOnlyCollection<T>` sounds good :+1:

File: AtomEventStore/TypeResolutionEntry.cs (hunk 7 above)
Comment (ploeh) [inline AtomEventStore/TypeResolutionEntry.cs]: `ResolvedType` sounds good :+1:

File: AtomEventStore/TypeResolutionTable.cs (hunk 9 above)
Comment (moodmosaic) [inline AtomEventStore/TypeResolutionTable.cs]: Yes, and then it can throw an `InvalidOperationException`. (I thought that `ArgumentException` was better...)

Comment (ploeh) [inline AtomEventStore/TypeResolutionTable.cs]: I think `ArgumentException` makes more sense as well.

I wonder if we're talking past each other...? Why do you need a `try`/`catch` block to handle the case where the entry isn't there?

Comment (moodmosaic) [inline AtomEventStore/TypeResolutionTable.cs]: Perhaps some defensive coding is warranted and manually throw an `ArgumentException`. However, as far as I can tell, we are going to iterate the sequence twice.

That's why I thought that `Enumerable.Single` would be fine:
- If the entry is there, return it 
- If the entry doesn't exists, throw
- If more than one entries found, throw

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 4 above)
Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs:64]: How about this:

```
[Theory, AutoAtomData]
public void ResolveThrowsWhenInputCanNotBeMappedToProperType(
    TypeResolutionEntry notMapped,
    [Frozen]TypeResolutionEntry[] dummyInjectedIntoSut,
    [FavorArrays]TypeResolutionTable sut)
{
    Assert.Throws<ArgumentException>(() =>
        sut.Resolve(notMapped.LocalName, notMapped.XmlNamespace));
}
```

Now `notMapped` comes before `[Frozen]` and unless `TypeResolutionEntry` is `[Frozen]` too, it shouldn't be in `entries`.

An alternative is to manually supply the value for `notMapped` in the test...

File: AtomEventStore/TypeResolutionTable.cs (hunk 9 above)
Comment (ploeh) [inline AtomEventStore/TypeResolutionTable.cs]: Ah, I see, you're also testing for duplicate values...

Are duplicate values legal?

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 4 above)
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs:64]: Please forget about what you know about AutoFixture's data generation algorithms for a while:

Tests can be organised into _Arrange Act Assert_ phases. Another way to express that is _Given When Then_. What should be a _Given_ in this test case? In other words: what's the precondition for this test case?

Another, unrelated question is: what does freezing the entries accomplish?

File: AtomEventStore/TypeResolutionTable.cs (hunk 9 above)
Comment (moodmosaic) [inline AtomEventStore/TypeResolutionTable.cs]: They are, unless we require a `HashSet<T>` instead of an array in the constructor.

(Although without overriding `Equals` in `TypeResolutionEntry`, `HashSet<T>` might not work correctly.)

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 4 above)
Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs:64]: OK, I must admit that I've been taking into account how AutoFixture works by default. Obviously that can lead to erratic tests, specially if the default AutoFixture's behaviour change.

What we need here is:
- The SUT
- A `notMapped` entry, that doesn't exists into the SUT's Entries.

That should look better:

``` c#
[Theory, AutoAtomData]
public void ResolveThrowsWhenInputCanNotBeMappedToProperType(
    TypeResolutionTable sut)
{
    var dummyType = typeof(Version);
    var notMapped = 
        new TypeResolutionEntry("not", "mapped", dummyType);
    Assert.Throws<ArgumentException>(() =>
        sut.Resolve(notMapped.LocalName, notMapped.XmlNamespace));
}
```

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 6 above)
Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs:100]: There isn't so I'm going to add these. Thank you!

File: AtomEventStore/TypeResolutionTable.cs (hunk 9 above)
Comment (ploeh) [inline AtomEventStore/TypeResolutionTable.cs]: Requiring a set instead of an collection sounds like a good idea, but most sets I've seen in .NET are mutable. Are you aware of any immutable set types in .NET (apart from in FSharp.Core, which I don't think we should pull in)?

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 4 above)
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs:64]: Still, how do you _know_ that, by some weird coincidence (or because of an AutoFixture customization), one of the injected entries _doesn't_ contain "not" "mapped"?

Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs:64]: I don't, unless I do this in the _Arrange_ phase:

``` c#
Assert.False(sut.Entries.Any(x => 
    x.LocalName == notMapped.LocalName && 
    x.XmlNamespace == notMapped.XmlNamespace));
```

File: AtomEventStore/TypeResolutionTable.cs (hunk 9 above)
Comment (moodmosaic) [inline AtomEventStore/TypeResolutionTable.cs]: I just found out that there's [`ImmutableHashSet<T>`](http://msdn.microsoft.com/en-us/library/dn467171.aspx), only available via [Microsoft.Bcl.Immutable](https://www.nuget.org/packages/Microsoft.Bcl.Immutable) NuGet package. However, I haven't used it before and I'm not aware of its performance characteristics...

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 4 above)
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs:64]: Bingo :) That's called a _Guard Assertion_.

Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs:64]: OK! :)

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 1 above)
Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Thank you for the hint! Done in 01826421324bcc5c5f06cd1d6b5f12b4cc5bb621.

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 2 above)
Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Thank you for the hint! Done in 3127a23b751596f68a45c47b6500cf4fc7f21a15.

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 3 above)
Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Thank you for the hint! Done in 6666e618d00447a17d94dc03c5381df2d4de2a2f.

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 4 above)
Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs:64]: Done in cd3e7aa1e28bde697fa212372828900ae7d3eca4.

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 6 above)
Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs:100]: Done in f4a21aa4e6a7c7d1461a08dcc0d7bb253735ae5e and 8f3f4655c4548401d399131bb268f7df5590540f.

File: AtomEventStore/TypeResolutionEntry.cs (hunk 7 above)
Comment (moodmosaic) [inline AtomEventStore/TypeResolutionEntry.cs]: Done in 61b72117fe006875a7bd564a08ff3bc342166649.

File: AtomEventStore/TypeResolutionTable.cs (hunk 8 above)
Comment (moodmosaic) [inline AtomEventStore/TypeResolutionTable.cs:33]: Done in 9b241288e3c7967e9849c25f70d1cf4a9c580c5e.

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 5 above)
Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Done in cd3e7aa1e28bde697fa212372828900ae7d3eca4.

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+            [Frozen]IReadOnlyCollection<TypeResolutionEntry> expected,
+            [FavorEnumerables]TypeResolutionTable sut)
+        {
+            var actual = sut.Entries;
+            Assert.True(expected.SequenceEqual(actual));
+        }
+
+        [Theory, AutoAtomData]
+        public void ResolveReturnsCorrectResult(
+            [FavorArrays]TypeResolutionTable sut)
```
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Why have `[FavorArrays]` here?

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+            Assert.Equal(expected, actual);
+        }
+
+        [Theory, AutoAtomData]
+        public void ResolveThrowsWhenInputCanNotBeMappedToProperType(
+            TypeResolutionTable sut)
+        {
+            var dummyType = typeof(Version);
+            var notMapped =
+                new TypeResolutionEntry("not", "mapped", dummyType);
```
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Why not just ask for `notMapped` via a method argument?

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+                x.LocalName == notMapped.LocalName &&
+                x.XmlNamespace == notMapped.XmlNamespace));
+
+            Assert.Throws<ArgumentException>(() =>
+                sut.Resolve(notMapped.LocalName, notMapped.XmlNamespace));
+        }
+
+        [Theory, AutoAtomData]
+        public void ResolveThrowsForUnmappedLocalNameAndMappedXmlNamespace(
+            [FavorArrays]TypeResolutionTable sut)
```
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Why have `[FavorArrays]` here?

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+                x.XmlNamespace == notMapped.XmlNamespace));
+            var mapped = sut.Entries.ToArray().PickRandom();
+
+            Assert.Throws<ArgumentException>(() =>
+                sut.Resolve(notMapped.LocalName, mapped.XmlNamespace));
+        }
+
+        [Theory, AutoAtomData]
+        public void ResolveThrowsForMappedLocalNameAndUnmappedXmlNamespace(
+            [FavorArrays]TypeResolutionTable sut)
```
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Why have `[FavorArrays]` here?

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+                sut.Resolve(notMapped.LocalName, notMapped.XmlNamespace));
+        }
+
+        [Theory, AutoAtomData]
+        public void ResolveThrowsForUnmappedLocalNameAndMappedXmlNamespace(
+            [FavorArrays]TypeResolutionTable sut)
+        {
+            var dummyType = typeof(Version);
+            var notMapped =
+                new TypeResolutionEntry("not", "mapped", dummyType);
```
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Why not just ask for `notMapped` via a method argument?

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+                sut.Resolve(notMapped.LocalName, mapped.XmlNamespace));
+        }
+
+        [Theory, AutoAtomData]
+        public void ResolveThrowsForMappedLocalNameAndUnmappedXmlNamespace(
+            [FavorArrays]TypeResolutionTable sut)
+        {
+            var dummyType = typeof(Version);
+            var notMapped =
+                new TypeResolutionEntry("not", "mapped", dummyType);
```
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Why not just ask for `notMapped` via a method argument?

File: AtomEventStore/TypeResolutionTable.cs [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+
+            try
+            {
+                return this.entries
+                    .Single(x =>
+                        x.LocalName == localName &&
+                        x.XmlNamespace == xmlNamespace)
+                    .ResolvedType;
+            }
+            catch (InvalidOperationException e)
```
Comment (ploeh) [inline AtomEventStore/TypeResolutionTable.cs]: I still don't understand why a `try`/`catch` block is necessary here. Why not invoke `SingleOrDefault` and check if the result is `null`?

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 10 above)
Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Because otherwise the `TypeResolutionTable(params TypeResolutionEntry[] entries)` is called but the entries are 0.

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 12 above)
Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Because otherwise the `TypeResolutionTable(params TypeResolutionEntry[] entries)` is called but the entries are 0.

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 13 above)
Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Because otherwise the `TypeResolutionTable(params TypeResolutionEntry[] entries)` is called but the entries are 0.

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 11 above)
Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Indeed. Done now in 1758161db19cebad5d5c39d585500b6a82e85990.

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 14 above)
Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Indeed. Done now in 1758161db19cebad5d5c39d585500b6a82e85990.

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 15 above)
Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Indeed. Done now in 1758161db19cebad5d5c39d585500b6a82e85990.

File: AtomEventStore/TypeResolutionTable.cs (hunk 16 above)
Comment (moodmosaic) [inline AtomEventStore/TypeResolutionTable.cs]: Indeed, all I needed is `SingleOrDefault`. Done in 3bb1136e803390efcd13da9e78c53436c699d7d2.

File: AtomEventStore.UnitTests/TypeResolutionTableTests.cs (hunk 10 above)
Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Oh, is that because AutoFixture doesn't understand `IReadOnlyCollection<T>`?

Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Yes, AFAICT, because AutoFixture targets .NET 4.0 and `IReadOnlyCollection<T>` is defined in .NET 4.5.

Comment (ploeh) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Wouldn't it be possible to customize AutoFixture in this particular code base to handle `IReadOnlyCollection<T>`?

Comment (moodmosaic) [inline AtomEventStore.UnitTests/TypeResolutionTableTests.cs]: Great idea! Done in c6d0ec362a557896afbd976ca4f0f4427bbc2842.

**Comment by ploeh** [comment]
Thank you for your contribution! It's now live as [AtomEventStore 1.3.0](https://www.nuget.org/packages/AtomEventStore/1.3.0).

**Comment by moodmosaic** [comment]
Thank you for your help and feedback!
